### PR TITLE
BlobsByRoot RPC

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

BlobsByRoot p2p RPC (spec: https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md)

### CODE REVIEW

File: beacon-chain/p2p/rpc_topic_mappings.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ const BlobsSidecarsByRangeMessageName = "/blobs_sidecars_by_range"
 // BeaconBlockAndBlobsSidecarByRootName is a topic for fetching beacon blocks and sidecar blobs by root.
 const BeaconBlockAndBlobsSidecarByRootName = "/beacon_block_and_blobs_sidecar_by_root"
```
Comment (kasey) [inline beacon-chain/p2p/rpc_topic_mappings.go:45]: comment for this is stranded in the other branch

File: beacon-chain/p2p/testing/fuzz_p2p.go [hunk 2]
```diff
@@ -138,6 +138,11 @@ func (_ *FakeP2P) BroadcastAttestation(_ context.Context, _ uint64, _ *ethpb.Att
 	return nil
 }
 
+// BroadcastBlob -- fake.
+func (p *FakeP2P) BroadcastBlob(ctx context.Context, subnet uint64, blobSidecar *ethpb.SignedBlobSidecar) error {
+	return nil
+}
+
```
Comment (kasey) [inline beacon-chain/p2p/testing/fuzz_p2p.go:145]: unrelated to this PR, I need to add this to run tests over the sync package.

File: beacon-chain/p2p/testing/mock_broadcaster.go [hunk 3]
```diff
@@ -33,3 +33,9 @@ func (m *MockBroadcaster) BroadcastSyncCommitteeMessage(_ context.Context, _ uin
 	m.BroadcastCalled = true
 	return nil
 }
+
+// BroadcastSyncCommitteeMessage records a broadcast occurred.
+func (m *MockBroadcaster) BroadcastBlob(ctx context.Context, subnet uint64, blob *ethpb.SignedBlobSidecar) error {
+	m.BroadcastCalled = true
+	return nil
+}
```
Comment (kasey) [inline beacon-chain/p2p/testing/mock_broadcaster.go:41]: same as above, just adding to get sync package tests to build.

File: beacon-chain/p2p/testing/p2p.go [hunk 4]
```diff
@@ -170,6 +170,11 @@ func (p *TestP2P) BroadcastAttestation(_ context.Context, _ uint64, _ *ethpb.Att
 	return nil
 }
 
+func (p *TestP2P) BroadcastBlob(ctx context.Context, subnet uint64, blobSidecar *ethpb.SignedBlobSidecar) error {
+	p.BroadcastCalled = true
+	return nil
+}
+
```
Comment (kasey) [inline beacon-chain/p2p/testing/p2p.go:177]: same as above, just adding to get sync package tests to build.

File: beacon-chain/p2p/types/rpc_errors.go [hunk 5]
```diff
@@ -12,4 +12,5 @@ var (
 	ErrRateLimited            = errors.New("rate limited")
 	ErrIODeadline             = errors.New("i/o deadline exceeded")
 	ErrInvalidRequest         = errors.New("invalid range, step or count")
+	ErrBlobLTMinRequest       = errors.New("blob slot < minimum_request_epoch")
```
Comment (kasey) [inline beacon-chain/p2p/types/rpc_errors.go:15]: I was thinknig "less than min request" which felt clear from the actual error text but this var name does look sorta weird to me. With fresh eyes I'm wondering "what is LTM and what does 'LTM in request' mean"?

File: beacon-chain/sync/database.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+type BlobSidecarWriter interface {
+	WriteBlobSidecar(blockRoot [32]byte, index uint64, sidecar *eth.BlobSidecar) error
+}
+
+type BlobDB interface {
+	BlobSidecarGetter
+	BlobSidecarWriter
+}
```
Comment (kasey) [inline beacon-chain/sync/database.go:16]: there is no concrete implementation of this interface, so while beacon-chain currently builds, nothing satisfying this interface is given to the sync service at init time, so methods that use this (like blobs by root) will throw a nil pointer exception.

File: beacon-chain/sync/rpc_blob_sidecars_by_root.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+			s.writeErrorResponseToStream(responseCodeServerError, types.ErrGeneric.Error(), stream)
+			tracing.AnnotateError(span, chunkErr)
+			return chunkErr
+		}
+		s.rateLimiter.add(stream, 1)
+	}
+	return nil
+}
+
+func (s *Service) sendBlocksAndSidecarsRequest(ctx context.Context, blockRoots *types.BeaconBlockByRootsReq, id peer.ID) error {
```
Comment (kasey) [inline beacon-chain/sync/rpc_blob_sidecars_by_root.go:82]: this is just some sloppypasta right here... :scissors:

File: beacon-chain/sync/rpc_blob_sidecars_by_root.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/prysmaticlabs/prysm/v3/monitoring/tracing"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v3/time/slots"
+	"go.opencensus.io/trace"
+)
+
+func minimumRequestEpoch(finalized, current primitives.Epoch) primitives.Epoch {
+	// max(finalized_epoch, current_epoch - MIN_EPOCHS_FOR_BLOB_SIDECARS_REQUESTS, DENEB_FORK_EPOCH)
+	denebFork := params.BeaconConfig().DenebForkEpoch
+	reqWindow := current - params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
```
Comment (terencechain) [inline beacon-chain/sync/rpc_blob_sidecars_by_root.go:25]: Underflow if `MinEpochsForBlobsSidecarsRequest > current`?

File: beacon-chain/sync/rpc_blob_sidecars_by_root.go (hunk 7 above)
Comment (terencechain) [reply beacon-chain/sync/rpc_blob_sidecars_by_root.go:82]: all good. We'll need to change it later anyway to remove `SignedBeaconBlockAndBlobsSidecar`

**Review by terencechain — COMMENTED** [review]
Nice work!

File: beacon-chain/sync/rpc_blob_sidecars_by_root.go (hunk 8 above)
Comment (kasey) [reply beacon-chain/sync/rpc_blob_sidecars_by_root.go:25]: good catch

**Pull request merged**